superduperdb/components/graph.py

```diff
@@ -0,0 +1,267 @@
+from functools import cached_property
+from inspect import signature
+
+import networkx as nx
+
+from superduperdb.base.artifact import Artifact
+from superduperdb.components.component import Component
+from superduperdb.components.model import Model
+from superduperdb.ext.torch.model import TorchModel
+
+if t.TYPE_CHECKING:
+    from superduperdb.base.datalayer import Datalayer
+
+
+class KeyStore:
+    def __init__(self, model, params):
+        self.model = model
+        self._params = params
+        for param in params:
+            setattr(self, param, f'{model}.{param}')
+
+    def __len__(self):
+        return len(self._params)
+
+    @property
+    def params(self):
+        return self._params
+
+
+class RootModel:
+    def __init__(self):
+        self.input = KeyStore('root', [])
+
+    def predict(self, x):
+        return x
+
+
+class GraphModel:
+    def __init__(self, model: t.Union['Model', 'TorchModel']):
+        self.identifier = model.identifier
+        self.output = f'{model.identifier}.output'
+
+        predict_method = model.predict_method
+        if isinstance(model, TorchModel):
+            predict_method = model.forward_method
+
+        assert isinstance(model.object, Artifact)
+        self.object = model.object.artifact
+
+        if predict_method:
+            self.predict_method = getattr(self.object, predict_method)
+        else:
+            self.predict_method = self.object
+
+        params = self.set_input()
+        self.input = KeyStore(model.identifier, params)
+
+    def set_input(self):
+        sig = signature(self.predict_method)
+        return list(sig.parameters.keys())
+
+    def predict(self, *args, **kwargs):
+        return self.predict_method(*args, **kwargs)
+
+
+class Node:
+    def __init__(self, name, db=None):
+        self.name = name
+
+        if name == Graph._input:
+            self.model = RootModel()
+        else:
+            self.model = GraphModel(db.load('model', name))
+        self._output = None
+
+    @cached_property
+    def params(self):
+        return self.model.input.params
+
+    @cached_property
+    def required_input(self):
+        return len(self.params)
+
+    @property
+    def output(self):
+        return self._output
+
+    @output.setter
+    def output(self, output):
+        if self._output:
+            raise ValueError('Not allowed to set output of a graph node')
+        self._output = output
+
+    def predict(self, *args, **kwargs):
+        self.output = self.model.predict(*args, **kwargs)
+        return self.output
+
+
+@dc.dataclass(kw_only=True)
+class Graph(Component):
+    db: 'Datalayer'
+    _DEFAULT_ARG_WEIGHT: t.ClassVar[str] = '_base'
+    _input: t.ClassVar[str] = '_Graph_input'
+
+    def __post_init__(self):
+        self.G = nx.DiGraph()
+        self._key_store = {}
+        self.nodes = {}
+        self._node_output_cache = {}
+
+    def connect(
+        self,
+        u: t.Union[str, Component],
```

Comment:
Why `Component`?

Reply:
because it can be either Graph, Model, or str 
Hence str, Component 
I can't use` Graph` for annotation here